dask_expr/_shuffle.py

```diff
@@ -555,7 +571,8 @@ def _layer(self):
         )
 
         dsk = {}
-        token = self._name.split("-")[-1]
+        # Ensure that shuffles with different branch_ids have the same barrier
+        token = self._dep_name.split("-")[-1]
```

Comment:
From a conversation with @fjetter: This is severely leaking internals of the P2P implementation.